### ISSUE TITLE
pdns_auth server only allows hmac-md5 algorithm with dnsupdate, blocking all others

### ISSUE THREAD

**Issue by PenelopeFudd** [PowerDNS/pdns#12182]
- Program: Authoritative
 - Issue type: Bug report

### Short description
Only TSIG keys generated with hmac-md5 work with dnsupdate, the others (hmac-sha1, hmac-sha224, hmac-sha256, hmac-sha384, and hmac-sha512) don't work, giving the error: 
```
; TSIG error with server: expected a TSIG or SIG(0)
update failed: REFUSED
```

### Environment

 - Operating system: Ubuntu 20.04
 - Software version: pdns-server 4.7.2-1pdns.focal amd64 
 - Software source: PowerDNS repository

### Steps to reproduce

1. Generate a tsig key: `pdnsutil generate-tsig-key test hmac-sha512`
2. Add the tsig key to a zone: `pdnsutil set-meta example.com TSIG-ALLOW-DSUPDATE test` 
3. Use the key: `echo 'server 127.0.0.4 53,zone example.com,update add test1.example.com 3600 TXT "test hmac-sha512",key test FDtcuNYhWctw2RPCZhYmOUgLqeIkxuUlUwR8h9ssCb8AS9o+VErySf2+0p8OR3lB5BY1Sm4FdtiIsXmbDWfY/A==,send,' | tr ',' '\n' | nsupdate`

### Expected behaviour
It should create a TXT record named 'test1.example.com' with the value "test hmac-sha512"

### Actual behaviour
```
; TSIG error with server: expected a TSIG or SIG(0)
update failed: REFUSED
```
### Other information
```
cat <<'END' > nsup.sh; chmod a+rx nsup.sh
#!/bin/bash

function die() { echo "${0}:${BASH_LINENO[0]}: $*" >&2; exit 1; }
function info() { echo "${0}:${BASH_LINENO[0]}: $*" >&2; }
function pdnsutil() { command pdnsutil --config-dir=/etc/powerdns/pdns_server-external "$@"; }

if [[ "$1" == "" ]]; then
  echo "Usage: ${0##*/} [-a hmac] (add|delete) \$zone \$name (A|TXT|CNAME|PTR|NS|SOA|etc) \$data"
  echo "  When adding, all the fields are needed; when deleting, only the zone and name are."
  exit 1
fi

# Selecting the algorithm:
ALGORITHM="hmac-md5"
if [[ "$1" == "-a" ]]; then shift; ALGORITHM="$1"; shift; fi

if [[ ! "$ALGORITHM" =~ ^(hmac-md5|hmac-sha1|hmac-sha224|hmac-sha256|hmac-sha384|hmac-sha512)$ ]]; then
  die "Algorithm $ALGORITHM is not one of (hmac-md5|hmac-sha1|hmac-sha224|hmac-sha256|hmac-sha384|hmac-sha512)"
fi

# Positional parameters:
operation="$1"; shift
     zone="$1"; shift
     name="$1"; shift
   rrtype="$1"; shift
     data="$*"

if [[ ! "$operation" =~ ^(add|del|delete)$ ]]; then
  die "The operation must be add or delete, not $operation"
fi

if [[ ! "$zone" =~ ^[a-zA-Z0-9][a-zA-Z0-9.-]*$ ]]; then
  die "The zone must be a domain name"
fi

if [[ ! "$name" =~ ^[a-zA-Z0-9][a-zA-Z0-9.-]*[.]${zone}$ ]]; then
  die "The zone must be a domain name ending with '$zone', not '$name'"
fi

if [[ "$operation" == add ]]; then

  if [[ ! "${rrtype^^}" =~ ^(A|AAAA|CAA|CNAME|DNSKEY|DS|HTTPS|MX|NS|PTR|SOA|SRV|TXT|TYPE10)$ ]]; then
    die "The type must be one of these: (A|AAAA|CAA|CNAME|DNSKEY|DS|HTTPS|MX|NS|PTR|SOA|SRV|TXT|TYPE10)"
  fi

  if [[ "$data" == "" ]]; then
    die "The data can't be blank."
  fi

fi

# Construct the arguments for the update command:
args=( "$operation" "$name" 3600 )
if [[ "$rrtype" != "" ]]; then args+=( "$rrtype" ); fi
if [[ "$data" != "" ]]; then 
  if [[ "$data" =~ ^[0-9][0-9]*[.][0-9][0-9]*[.][0-9][0-9]*[.][0-9][0-9]*$ ]]; then
    args+=( "$data" )
  else
    args+=( "\"$data\"" )
  fi
fi

if [[ "$EUID" != 0 ]]; then die "This script needs root access."; fi

a=$(pdnsutil generate-tsig-key test "$ALGORITHM") \
  || die "'pdnsutil generate-tsig-key test $ALGORITHM' failed"
echo "$a"

pdnsutil set-meta netskrt.org TSIG-ALLOW-DNSUPDATE test \
  || die "'pdnsutil set-meta netskrt.org TSIG-ALLOW-DNSUPDATE test' failed"

keytext=$(echo "$a" | sed 's/.* //')

message=$(cat <<EOF
  server 127.0.0.4 53
  zone $zone
  update ${args[@]}
  key test $keytext
  send
EOF
)

if ! echo "$message" | nsupdate; then
  echo ""
  info "nsupdate failed with this input:
=====
$message
====="
fi

pdnsutil delete-tsig-key test || die "'pdnsutil delete-tsig-key test' failed"

nslookup -q="$rrtype" "$name" 127.0.0.4 | tail -n +3
END

for n in hmac-md5 hmac-sha1 hmac-sha224 hmac-sha256 hmac-sha384 and hmac-sha512; do
  nsup.sh -a "$n" add example.com test1.example.com TXT "test $n" >& /dev/null
done

nslookup -q=txt test1.example.com 127.0.0.4

Server:		127.0.0.4
Address:	127.0.0.4#53

test1.netskrt.org	text = "test hmac-md5"
```
In other words, the only algorithm that worked was `hmac-md5`, the others failed.

Coincidentally, the [dnsupdate example](https://doc.powerdns.com/authoritative/dnsupdate.html#tsig-allow-dnsupdate) demonstrates using the hmac-md5 algorithm, but not the others, which may explain why this hasn't been detected before.

**Comment by cmouse** [comment]
I can't figure out from your script if you are using [hmac:]keyname:secret as key format.

for sha256, this would be hmac-sha256:test:secret

**Comment by Habbie** [comment]
Just for completeness, what version of auth please?

**Comment by PenelopeFudd** [comment]
The key format for nsupdate is: 
```
  server 127.0.0.4 53
  zone example.com
  update add test1.example.com 3600 txt "test99 hmac-sha512"
  key test HnKV7gHiKGi5virOzxW0Y+NwhKVTxwjtd0aG0Z6gmME2DjBmDmqjNmfYxAGAGplv+3YWovkxwhH3pmHTkMJiiw==
  send
```
We're using TSIG auth?  I'm unclear if that's the answer you're looking for.

**Comment by PenelopeFudd** [comment]
The nsupdate example given [here](https://doc.powerdns.com/authoritative/dnsupdate.html#tsig-allow-dnsupdate) is: 
nsupdate <<!
server <ip> <port>
zone example.org
update add test1.example.org 3600 A 203.0.113.1
key test kp4/24gyYsEzbuTVJRUMoqGFmN3LYgVDzJ/3oRSP7ys=
send
!

**Comment by PenelopeFudd** [comment]
Would I be correct in inferring that `nsupdate` defaults to hmac-md5 unless explicitly stated?

**Comment by cmouse** [comment]
Yep. You can try `nsupdate -k hmac-sha512:test:HnKV7gHiKGi5virOzxW0Y+NwhKVTxwjtd0aG0Z6gmME2DjBmDmqjNmfYxAGAGplv+3YWovkxwhH3pmHTkMJiiw==`

**Comment by PenelopeFudd** [comment]
It doesn't seem to like it:
```
# nsupdate -y hmac-sha512:test99:mZws+E6Ho14a6uGJZAbvtXgTvrLk/d9ZuB28Ng3UVijhdXY+dYLUnAIYofJLP/f81X66QPGsuGM/5VJSZZewRg==
 
> server 127.0.0.4 53
> zone example.com
> update add test1.example.com 3600 TXT "trying -y option on cli"
> send
; TSIG error with server: expected a TSIG or SIG(0)
update failed: REFUSED
```

**Comment by cmouse** [comment]
And it seems that you can use `key hmac-sha512:test HnKV7gHiKGi5virOzxW0Y+NwhKVTxwjtd0aG0Z6gmME2DjBmDmqjNmfYxAGAGplv+3YWovkxwhH3pmHTkMJiiw==`

with cli.

did you check the server logs?

**Comment by PenelopeFudd** [comment]
```
# pde list-tsig-keys | grep test99
test99. hmac-md5. CVLBvwi/WxuKO9EjQ8z32lsy9c0W5ynU4UDJAnDXZEY=
test99. hmac-sha512. mZws+E6Ho14a6uGJZAbvtXgTvrLk/d9ZuB28Ng3UVijhdXY+dYLUnAIYofJLP/f81X66QPGsuGM/5VJSZZewRg==

# nsupdate -y hmac-sha512:test99:mZws+E6Ho14a6uGJZAbvtXgTvrLk/d9ZuB28Ng3UVijhdXY+dYLUnAIYofJLP/f81X66QPGsuGM/5VJSZZewRg==
> server 127.0.0.4 53
> zone example.com
> update add test1.example.com 3600 TXT "test again"
> send
; TSIG error with server: expected a TSIG or SIG(0)
update failed: REFUSED
```

**Comment by cmouse** [comment]
did you check server logs?

**Comment by PenelopeFudd** [comment]
```
Nov 10 00:23:34 node3 pdns_server-external[2726510]: Packet for 'example.com' denied: Signature with TSIG key 'test99' does not match the expected algorithm (hmac-md5 / hmac-sha512)
Nov 10 00:23:34 node3 pdns_server-external[2726510]: Received a TSIG signed message with a non-validating key
```

**Comment by PenelopeFudd** [comment]
Oh, wait, bad example, I used the older test program that didn't use `-y hmac-sha512:test:.......` on the command line.

Redoing test....

**Comment by PenelopeFudd** [comment]
Ok, I suppose I should have read the nsupdate man page:

```
          -y [hmac:]keyname:secret
           Literal TSIG authentication key.  keyname is the name of the key, and secret is the base64 encoded shared secret.  hmac is the name of the key algorithm; valid
           choices are hmac-md5, hmac-sha1, hmac-sha224, hmac-sha256, hmac-sha384, or hmac-sha512. If hmac is not specified, the default is hmac-md5 or if MD5 was disabled
           hmac-sha256.
```

Nsupdate defaults to hmac-md5 if not specified.  D'oh!

Small request: could you enhance the example code in the documentation to make the algorithm explicit?  :-)
https://doc.powerdns.com/authoritative/dnsupdate.html#tsig-allow-dnsupdate

Thanks!

**Comment by Habbie** [comment]
> Small request: could you enhance the example code in the documentation to make the algorithm explicit? :-)

We can, but in a closed ticket, we'll forget. Can you make another ticket, or perhaps a PR? Thanks!

**Comment by PenelopeFudd** [comment]
PR created, and ticket reopened for good measure!  :-)

**Comment by Habbie** [comment]
PR merged (some time ago ;) )

closing this
